src/macaron/slsa_analyzer/build_tool/base_build_tool.py

```diff
@@ -107,12 +111,57 @@ def prepare_config_files(self, wrapper_path: str, build_dir: str) -> bool:
         bool
             True if succeed else False.
         """
-        raise NotImplementedError
 
     @abstractmethod
     def load_defaults(self) -> None:
         """Load the default values from defaults.ini."""
-        raise NotImplementedError
+
+    @abstractmethod
+    def get_dep_analyzer(self, repo_path: str) -> DependencyAnalyzer:
+        """Create a DependencyAnalyzer for the build tool.
+
+        Parameters
+        ----------
+        repo_path: str
+            The path to the target repo.
+
+        Returns
+        -------
+        DependencyAnalyzer
+            The DependencyAnalyzer object.
+        """
+
+    def get_build_dirs(self, repo_path: str) -> Iterable[Path]:
+        """Find directories in the repository that have their own build scripts.
+
+        This is especially important for applications that consist of multiple services.
+
+        Parameters
+        ----------
+        repo_path: str
+            The path to the target repo.
+
+        Yields
+        ------
+        Path
+            The relative paths from the repo path that contain build scripts.
+        """
+        config_paths: set[str] = set()
+        for build_cfg in self.build_configs:
+            config_paths.update(glob.glob(os.path.join(repo_path, "**", build_cfg), recursive=True))
+
+        list_iter = iter(sorted(config_paths, key=lambda x: (str(Path(x).parent), len(Path(x).parts))))
+        try:
+            prefix = next(list_iter)
+            yield Path(prefix).parent.relative_to(repo_path)
```

Comment:
If the elements in `list_iter` are already the path to the parent dir of build config files discovered in a repository, why do we need to find the parent here again?

Reply:
The items in `list_iter` are the full path to the config files. I think `prefix` is a confusing variable name. I'll change it to `cfg_path`.